src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -731,55 +757,21 @@ void ProcessOverride (OverrideInformation overrideInformation)
 				ProcessVirtualMethod (method);
 		}
 
-		bool IsInterfaceOverrideThatDoesNotNeedMarked (OverrideInformation overrideInformation, bool isInstantiated)
-		{
-			if (!overrideInformation.IsOverrideOfInterfaceMember || isInstantiated)
-				return false;
-
-			// This is a static interface method and these checks should all be true
-			if (overrideInformation.Override.IsStatic && overrideInformation.Base.IsStatic && overrideInformation.Base.IsAbstract && !overrideInformation.Override.IsVirtual)
-				return false;
-
-			if (overrideInformation.MatchingInterfaceImplementation != null)
-				return !Annotations.IsMarked (overrideInformation.MatchingInterfaceImplementation);
-
-			var interfaceType = overrideInformation.InterfaceType;
-			var overrideDeclaringType = overrideInformation.Override.DeclaringType;
-
-			if (interfaceType == null || !IsInterfaceImplementationMarkedRecursively (overrideDeclaringType, interfaceType))
-				return true;
-
-			return false;
-		}
-
+		/// <summary>
+		/// Returns true if <paramref name="type"/> implements <paramref name="interfaceType"/> and the interface implementation is marked,
+		/// or if any marked interface implementations on <paramref name="type"/> are interfaces that implement <paramref name="interfaceType"/> and that interface implementation is marked
+		/// </summary>
 		bool IsInterfaceImplementationMarkedRecursively (TypeDefinition type, TypeDefinition interfaceType)
 		{
+			if (type.Equals (interfaceType))
+				return true;
 			if (type.HasInterfaces) {
 				foreach (var intf in type.Interfaces) {
 					TypeDefinition? resolvedInterface = Context.Resolve (intf.InterfaceType);
 					if (resolvedInterface == null)
 						continue;
 
-					if (Annotations.IsMarked (intf) && RequiresInterfaceRecursively (resolvedInterface, interfaceType))
-						return true;
-				}
-			}
-
-			return false;
-		}
-
-		bool RequiresInterfaceRecursively (TypeDefinition typeToExamine, TypeDefinition interfaceType)
-		{
-			if (typeToExamine == interfaceType)
-				return true;
-
-			if (typeToExamine.HasInterfaces) {
-				foreach (var iface in typeToExamine.Interfaces) {
-					var resolved = Context.TryResolve (iface.InterfaceType);
-					if (resolved == null)
-						continue;
-
-					if (RequiresInterfaceRecursively (resolved, interfaceType))
+					if (Annotations.IsMarked (intf) && IsInterfaceImplementationMarkedRecursively (resolvedInterface, interfaceType))
```

Comment:
There is a test case (OverrideRemoval.OverrideThatAlsoFulfilsInterface)[https://github.com/dotnet/linker/blob/main/test/Mono.Linker.Tests.Cases/Inheritance.AbstractClasses/NoKeptCtor/OverrideRemoval/OverrideThatAlsoFulfilsInterface.cs] that hits an issue when I removed the check (Foo.Method() should be removed but isn't). The original implementation checked the first level interface implementation then didn't check any others to get around this. I think this might be a bit conservative, and I think we could check for the interface implementation to be either marked OR the implementing type to be in a preserved scope to cover the scenario you mentioned. If it seems like there could still be a hole in that, I could switch it back to the original implementation.